tests/test_forward_ref.py

```diff
@@ -840,3 +859,95 @@ class Foo(BaseModel):
+    usage: `with pytest_raises_undefined_types_warning(class_name='Foobar', missing_class_name='UndefinedType'):`
+    """
+
+    return pytest.raises(
+        UserWarning,
+        match=re.escape(
+            # split this f-string into two lines to pass linting (line too long)
+            f'`{defining_class_name}` is not fully defined, you should define `{missing_type_name}`, '
+            f'then call `{defining_class_name}.model_rebuild()`',
+        ),
+    )
+
+
+def test_undefined_type_raises_warning(create_module):
+    """Test undefined_types_warning is raised when a model declares a variable of type that is missing."""
+
+    with pytest_raises_undefined_types_warning(defining_class_name='Foobar', missing_type_name='UndefinedType'):
+        create_module(
+            # language=Python
+            """
+from __future__ import annotations
+from pydantic import BaseModel
+
+class Foobar(BaseModel):
+    a: UndefinedType
+"""
+        )
+
+
+def test_undefined_type_raises_warning_if_forward_ref(create_module):
+    """Test undefined_types_warning is raised when a model declares a variable of type that is a ForwardRef."""
+
+    with pytest_raises_undefined_types_warning(defining_class_name='Foobar', missing_type_name='UndefinedType'):
+
+        @create_module
+        def module():
+            from typing import ForwardRef
+
+            from pydantic import BaseModel
+
+            UndefinedType = ForwardRef('UndefinedType')
+
+            class Foobar(BaseModel):
+                a: UndefinedType
+
+
+def test_disable_undefined_types_warning(create_module):
+    """Test suppressing undefined_types_warning via config \
+        when a model declares a variable of type that is missing."""
+
+    module = create_module(
+        # language=Python
+        """
+from __future__ import annotations
+from pydantic import BaseModel
+
+class Foobar(BaseModel):
+    a: UndefinedType
+    class Config:
+        undefined_types_warning = False
+"""
+    )
+    # Since test is about suppressing the undefined_types_warning,
+    # we should also verify that our model was at least partially built
+    assert module.Foobar.__pydantic_model_complete__ is False
+
+
+def test_disable_undefined_types_warning_if_missing_class_is_a_forwardref(create_module):
+    """Test suppressing undefined_types_warning via config \
+        when a model declares a variable of type that is a ForwardRef."""
+
+    @create_module
+    def module():
+        from typing import ForwardRef
+
+        from pydantic import BaseModel
+
+        UndefinedType = ForwardRef('UndefinedType')
+
+        class Foobar(BaseModel):
+            a: UndefinedType
+
+            class Config:
+                undefined_types_warning = False
+
+    # Since test is about suppressing the undefined_types_warning,
+    # we should also verify that our model was at least partially built
+    assert module.Foobar.__pydantic_model_complete__ is False
```

Comment:
@samuelcolvin @ybressler there are now 4 new unit tests which are basically a matrix of:

1) how to handle UndefinedType (import annotations vs import ForwardRef)
    a) use `from __future__ import annotations`
    b) use `UndefinedType = ForwardRef('UndefinedType')`
2) whether to suppress the undefined_types_warning (suppress warning or not)
    a) don't suppress it and then verify the expected warning and its detailed message
    b) do suppress it and then verify the model is partially built (without the warning being raised)

The test names reflect these 4 different possibilities of the two different parameters

Reply:
Great.

Reply:
Fantastic work!